optimization.ts

```diff
@@ -0,0 +1,433 @@
+import { BinOp, Parameter, Type, UniOp} from "./ast";
+import { Stmt, Expr, Value, VarInit, BasicBlock, Program, FunDef, Class } from "./ir";
+
+import { isTagBoolean, isTagNone, isTagId, isTagBigInt, isTagEqual, checkValueEquality, checkCompileValEquality, checkStmtEquality } from "./optimization_utils"; 
+
+type Env = {
+    vars: Map<string, compileVal>;
+}
+
+export type compileVal = {
+    tag: "nac"|"val"|"undef", value?: Value<any>;
+}
+
+export function optimizeValue(val: Value<any>, env: Env): Value<any>{
+    if (val.tag !== "id"){
+        return val;
+    }
+    if (env.vars.has(val.name)){
+        if (["nac", "undef"].includes(env.vars.get(val.name).tag))
+            return val;
+        val = env.vars.get(val.name).value;
+    }
+    return val;
+}
+
+export function evaluateBinOp(op: BinOp, leftVal: Value<any>, rightVal: Value<any>): Value<any>{
+    if([BinOp.Plus, BinOp.Minus,BinOp.IDiv,BinOp.Mul, BinOp.Gt, BinOp.Lt, BinOp.Gte, BinOp.Lte, BinOp.Mod].includes(op)){
+        if(!isTagBigInt(leftVal) || !isTagBigInt(rightVal))
+            throw new Error("Compiler Error");
```

Comment:
This error is interesting; I think that w/strings we may or may not run into some issues depending on how we define values. Worth thinking a little bit about how fallthrough cases like these will work in the presence of new types.

Reply:
Updated the error message (ideally, this should never be thrown if everything prior to that works fine - more to convince typescript of the types)